src/stim/circuit/gate_data.cc

```diff
@@ -81,25 +81,293 @@ Gate::Gate() : name(nullptr) {
 
 Gate::Gate(
     const char *name,
-    const char *best_inverse_name,
+    Gates gate_id,
+    Gates best_inverse_gate,
     uint8_t arg_count,
-    void (TableauSimulator::*tableau_simulator_function)(const OperationData &),
-    void (FrameSimulator::*frame_simulator_function)(const OperationData &),
-    void (ErrorAnalyzer::*hit_simulator_function)(const OperationData &),
-    void (SparseUnsignedRevFrameTracker::*sparse_unsigned_rev_frame_tracker_function)(const OperationData &),
     GateFlags flags,
     ExtraGateData (*extra_data_func)(void))
     : name(name),
-      tableau_simulator_function(tableau_simulator_function),
-      frame_simulator_function(frame_simulator_function),
-      reverse_error_analyzer_function(hit_simulator_function),
-      sparse_unsigned_rev_frame_tracker_function(sparse_unsigned_rev_frame_tracker_function),
       extra_data_func(extra_data_func),
       flags(flags),
       arg_count(arg_count),
       name_len((uint8_t)strlen(name)),
-      id(gate_name_to_id(name)),
-      best_candidate_inverse_id(gate_name_to_id(best_inverse_name)) {
+      id(static_cast<uint8_t>(gate_id)),
+      best_candidate_inverse_id(static_cast<uint8_t>(best_inverse_gate)) {
+}
+
+template <typename SIMULATOR>
+constexpr void GateVTable<SIMULATOR>::add_simulator_function_annotations() {
+    if constexpr (std::is_same<SIMULATOR, TableauSimulator>::value ||
+                  std::is_same<SIMULATOR, FrameSimulator>::value) {
+        add_simulator_function(Gates::DETECTOR, &SIMULATOR::I);
```

Comment:
Let's break the bad cyclic dependency where `gate_data` knows *anything* about any simulator, and instead rely on each simulator knowing about gate data. In particular, imagine adding a new simulator; it should be possible to do it with ideally zero touching of `gate_data.cc`. In fact, imagine adding the new simulator with no source code access (e.g. by linking to `libstim`). Make *that* work, and then use that for the simulators inside stim. Have each simulator define each gate's method, instead of each gate defining each simulator's method.

(It's okay that adding a gate requires updating every simulator in response. Bonus points if that causes a compile time failure, but realistically a unit test failure for that simulator is sufficient.)

Reply:
Good point. I took two steps to address this:

1. First in f12b298 I ensured that the function pointers had to be passed into the constructor of the VTable.
2. Then in fbfff42 I associated the VTables with each simulator via composition. It's a bit harder to reason about exactly how much happens at the compile stage after this step. I also wasn't able to keep the vtable const in the `SparseUnsignedRevFrameTracker`.

Reply:
That looks good. It's unfortunate that it makes the table at construction time, but that's not on the hot path and it avoids a lot of issues we'd run into when adding templates compared to trying to make the global.